### PR TITLE
#237 implemented validation of mandatory fields

### PR DESCRIPTION
**Pull request by rgraf**
Implemented validation of mandatory fields for Agent, VcardAddress and describing entity.

### CODE REVIEW

File: annotation-client/src/test/java/eu/europeana/annotation/client/integration/webanno/tag/SemanticTaggingTest.java [hunk 1]
```diff
@@ -85,4 +85,16 @@ public void createSemanticTagEntity() throws IOException, JsonParseException, Il
...
 
+	@Test
+	public void createSemanticTagAgentEntity() throws IOException, JsonParseException, IllegalAccessException,
+			IllegalArgumentException, InvocationTargetException {
+
+		Annotation storedAnno = createTag(SEMANTICTAG_AGENT_ENTITY, false, true);
+		log.info(storedAnno.getBody().getInternalType());
+		assertTrue(storedAnno.getMotivation().equals(MotivationTypes.TAGGING.name().toLowerCase()));
+		assertEquals(storedAnno.getBody().getInternalType(), BodyInternalTypes.AGENT.name());
+		EdmAgent agentBody = (EdmAgent) ((EdmAgentBody) storedAnno.getBody()).getAgent();
```
Comment (gsergiu) [inline annotation-client/src/test/java/eu/europeana/annotation/client/integration/webanno/tag/SemanticTaggingTest.java]: Do we support all fields from the given example?
We have to compare the body of the input with the body of the stored annotation similar to the the validateOutputAgainstInput(Annotation storedAnno, Annotation inputAnno)

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 2]
```diff
@@ -542,14 +549,17 @@ protected void validateEdmRights(Body body)
...
      */
-    private void validateTag(Annotation webAnnotation) throws ParamValidationException {
+    private void validateTag(Annotation webAnnotation) throws ParamValidationException, PropertyValidationException {
 	// webAnnotation.
 	Body body = webAnnotation.getBody();
 
 	// TODO: the body type shouldn't be null at this stage
 	if (body.getType() != null && body.getType().contains(WebAnnotationFields.SPECIFIC_RESOURCE)) {
 	    validateTagWithSpecificResource(body);
+	} else if (body.getType() != null && body.getType().contains(WebAnnotationFields.FULL_TEXT_RESOURCE)) {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: We don't have such a usecase, it must not be allowed to use fulltext resources when motivation is tagging. Please remove

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 3]
```diff
@@ -572,20 +595,56 @@ private void validateTag(Annotation webAnnotation) throws ParamValidationExcepti
...
+     * This method validates entity body.
      * 
      * @param body The entity body
      * @throws ParamValidationException
+	 * @throws PropertyValidationException 
      */
-    private void validateAgentBody(Body body) throws ParamValidationException {
-	if (!(body instanceof EdmAgentBody)) {
+    private void validateAgentBody(Body body) throws ParamValidationException, PropertyValidationException {
+		if (!(body instanceof EdmAgentBody)) {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: The check must not be based on the class, but on the body.getType, see validation of body type for transcriptions

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 4]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
 	if (StringUtils.isBlank(body.getEdmRights())) {
 	    throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
 		    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "transcription.body.edmRights" });
-	}
+    }
 	validateEdmRights(body);
     }
+    
+    private void validateTagWithFullTextResource(Body body) throws ParamValidationException, PropertyValidationException {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Please remove this method, we don't support such use case

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 5]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		// check mandatory fields
+    	validateSemanticTagVcardAddressBody(body);
+	
+		// check type
 	if (StringUtils.isBlank(body.getInternalType().toString())
-		|| !BodyInternalTypes.isVcardAddressTagBody(body.getInternalType()))
-	    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
-		    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
-		    new String[] { "tag.body.type", body.getType().toString() });
+			|| !BodyInternalTypes.isVcardAddressTagBody(body.getInternalType()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: This validation must not be done based on the internal type, but on the (external) getType, see validation for transcriptions

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 6]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+			    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
+			    new String[] { "tag.body.type", body.getType().toString() });
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Mandatory field message has only one parameter, please remove the second entry

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 7]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+     * Semantic tagging with vcard:Address type has mandatory fields:
+     * "Address", "streetAddress", "locality", "countryName"
+     * The "type" field should have value "Address"
+     * @param body
+     * @throws ParamValidationException
+	 * @throws PropertyValidationException 
+     */
+    private void validateSemanticTagVcardAddressBody(Body body) throws ParamValidationException, PropertyValidationException {
+		// check mandatory field address
+		if (!(body instanceof VcardAddressBody))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: the check must be implemented based on getType(), see transcriptions

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 8]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+     * @param body
+     * @throws ParamValidationException
+	 * @throws PropertyValidationException 
+     */
+    private void validateSemanticTagVcardAddressBody(Body body) throws ParamValidationException, PropertyValidationException {
+		// check mandatory field address
+		if (!(body instanceof VcardAddressBody)) 
+		    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+				    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
+				    new String[] { "tag.body.class", body.getClass().toString() });
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: remove second entry in the string list

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 9]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+
+		Address address = ((VcardAddressBody) body).getAddress();
+		
+		// check mandatory field streetAddress
+		if (Strings.isNullOrEmpty(address.getVcardStreetAddress()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Please remove the usage of Strings class and use StringUtils like the other methods

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 10]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.streetAddress" });
+
+		// check mandatory field locality
+		if (Strings.isNullOrEmpty(address.getVcardLocality()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 11]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.locality" });
+
+		// check mandatory field countryName
+		if (Strings.isNullOrEmpty(address.getVcardCountryName()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 12]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.countryName" });
+		
+		// check mandatory field type
+		if (body.getType() == null || Strings.isNullOrEmpty(body.getType().get(0))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils and also ensure that only one Type is used

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 13]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.locality" });
+
+		// check mandatory field countryName
+		if (Strings.isNullOrEmpty(address.getVcardCountryName()))
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.countryName" });
+		
+		// check mandatory field type
+		if (body.getType() == null || Strings.isNullOrEmpty(body.getType().get(0)) 
+				|| !body.getType().get(0).equals(WebAnnotationModelKeywords.CLASS_VCARD_ADDRESS))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Use ResourceTypes enumeration for verification of Types

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 14]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: When type is different than Address, a different message must be used, Invalid_property_value

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 15]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		
+    /**
+     * For describing annotations: "value" and "language" within the "body" are mandatory.
+     * @param body
+     * @throws ParamValidationException
+     */
+    private void validateTextualBody(Body body) throws ParamValidationException {
+
+		// check mandatory field value
+		if (Strings.isNullOrEmpty(body.getValue()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 16]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+			    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
+			    new String[] { "tag.body.value", body.getValue() });
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: remove second entry

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 17]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	
+		// check mandatory field language
+		if (Strings.isNullOrEmpty(body.getLanguage()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 18]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+			    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
+			    new String[] { "tag.body.language", body.getLanguage() });
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: remove second entry

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 19]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+
+		// check type
+		if (Strings.isNullOrEmpty(body.getInternalType().toString())
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 20]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+			|| !BodyInternalTypes.isTextualBody(body.getInternalType()))
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use getType and ResourceTypes for this check

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 21]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+		    throw new ParamValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+			    I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD,
+			    new String[] { "tag.body.type", body.getType().toString() });
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: remove second entry

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java [hunk 22]
```diff
@@ -671,24 +729,107 @@ private void validateTranscriptionWithFullTextResource(Body body)
...
+	        throw new PropertyValidationException(ParamValidationException.MESSAGE_MISSING_MANDATORY_FIELD,
+		        I18nConstants.MESSAGE_MISSING_MANDATORY_FIELD, new String[] { "tag.body.address.type" });
+    }
+		
+    /**
+     * For describing annotations: "value" and "language" within the "body" are mandatory.
+     * @param body
+     * @throws ParamValidationException
+     */
+    private void validateTextualBody(Body body) throws ParamValidationException {
```
Comment (gsergiu) [inline annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: add bollean param to indicate if the language is mandatory. There are usecases for which the language is not mandatory

File: annotation-client/src/test/java/eu/europeana/annotation/client/integration/webanno/tag/SemanticTaggingTest.java (hunk 1 above)
Comment (rgraf) [reply annotation-client/src/test/java/eu/europeana/annotation/client/integration/webanno/tag/SemanticTaggingTest.java]: Entsprechend dem Ticket ist nur skos:prefLabel mandatory. I added additional fields to check for particular example.

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 4 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Method removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 2 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: use case removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 3 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Replaced similar to transcriptions

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 7 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Implemented using getType() like in transcriptions

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 8 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 6 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Only one parameter remaining

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 5 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: this test is anyway covered in validateSemanticTagVcardAddressBody now

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 9 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: Replaced by StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 10 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: done with StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 11 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: done with StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 12 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: used StringUtils and that only one type is in use

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 21 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: second entry removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 20 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: used getType and ResourceTypes

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 19 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: used StringUtils

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 18 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: second entry removed

File: annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java (hunk 22 above)
Comment (rgraf) [reply annotation-web/src/main/java/eu/europeana/annotation/web/service/impl/AnnotationServiceImpl.java]: added boolean parameter for language

**Pull request merged**